pallets/erc20-xcm-bridge/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pallet that allow to transact erc20 tokens trought xcm directly.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+mod erc20_matcher;
+mod erc20_trap;
+mod errors;
+
+use frame_support::pallet;
+
+pub use erc20_trap::AssetTrapWrapper;
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use crate::erc20_matcher::*;
+	use crate::errors::*;
+	use ethereum_types::BigEndianHash;
+	use fp_evm::{ExitReason, ExitSucceed};
+	use frame_support::pallet_prelude::*;
+	use pallet_evm::Runner;
+	use sp_core::{H160, H256, U256};
+	use xcm::latest::{Error as XcmError, MultiAsset, MultiLocation, Result as XcmResult};
+	use xcm_executor::traits::{Convert, Error as MatchError};
+	use xcm_executor::Assets;
+
+	const ERC20_TRANSFER_GAS_LIMIT: u64 = 200_000;
+	const ERC20_TRANSFER_SELECTOR: [u8; 4] = [0xa9, 0x05, 0x9c, 0xbb];
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_evm::Config {
+		type AccountIdConverter: Convert<MultiLocation, H160>;
+		type Erc20MultilocationPrefix: Get<MultiLocation>;
+		type EvmRunner: Runner<Self>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn erc20_transfer(
+			erc20_contract_address: H160,
+			from: H160,
+			to: H160,
+			amount: U256,
+		) -> Result<(), Erc20TransferError> {
+			// ERC20.transfer method hash
+			let mut input = ERC20_TRANSFER_SELECTOR.to_vec();
```

Comment:
Will this result in multiple heap allocations? If so, consider pre-allocating the known amount

Reply:
Done